packages/passport-interface/src/util/StorageBackedMap.ts

```diff
@@ -0,0 +1,102 @@
+/**
+ * A JavaScript map which syncs its entries to local storage, and loads them
+ * during construction.
+ */
+export class StorageBackedMap<K, V> extends Map<K, V> {
+  // The local storage key
+  private readonly storageKey: string;
+  // Whether we're currently syncing to local storage
+  private syncing: boolean;
+
+  public constructor(storageKey: string) {
+    // Initializes the base map
+    super();
+    this.storageKey = storageKey;
+    this.syncing = false;
+
+    // Load data into the map
+    this.reloadFromStorage();
+
+    // Set up an event listener so that we get changes to local storage from
+    // other tabs
+    window.addEventListener("storage", (ev) => {
+      // key === null means that storage was cleared
+      // Probably this means that the user was logged out, so let's make sure
+      // we don't keep any credentials in memory
+      if (ev.key === null) {
+        super.clear();
+      } else if (ev.key === this.storageKey) {
+        this.reloadFromStorage();
+      }
+    });
+  }
+
+  // Queues a microtask to sync to local storage once the current event loop
+  // has finished processing
+  private queueSync() {
+    if (!this.syncing) {
+      this.syncing = true;
+      queueMicrotask(() => this.syncToStorage());
+    }
+  }
+
+  // Sync the map entries to local storage
+  private syncToStorage(): void {
+    const data = JSON.stringify(Array.from(this.entries()));
+    if (window.localStorage.getItem(this.storageKey) !== data) {
+      window.localStorage.setItem(this.storageKey, data);
+    }
+    this.syncing = false;
+  }
+
+  // Reloads data from storage, called in response to storage changes that
+  // come from other tabs.
+  private reloadFromStorage() {
+    const storageData = window.localStorage.getItem(this.storageKey);
+    let loadedData = [];
+    if (storageData) {
+      const parsed = JSON.parse(storageData);
```

Comment:
`JSON.parse` can throw an exception - i suggest surrounding this with a `try/catch` and handling the case where what's in localstorage is nonsense explicitly.